File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers: list, target: int):
    answer = 0

    def dfs(total, index):
        nonlocal answer
        
        if index == len(numbers):
            if total == target:
                answer += 1
        else:
            dfs(total + numbers[index], index + 1)
            dfs(total - numbers[index], index + 1)

    t = dfs(0, 0)
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {

    private static int N;
    private static int key;
    private static int answer;
    private static int[] nums;
    private static boolean[] visited;

    public int solution(int[] numbers, int target) {
        N = numbers.length;
        key = target;
        nums = numbers;
        visited = new boolean[N];
        dfs(0, 0);
        return answer;
    }

    private void dfs(int idx, int total) {
        if (idx >= N) {
            if (total == key) {
                answer++;
            }
            return;
        }
        dfs(idx + 1, total + nums[idx]);
        dfs(idx + 1, total - nums[idx]);
    }
}